src/Request.php

```diff
@@ -201,8 +201,15 @@ protected function get_validation_rules(): array {
 		$validation_rules = GraphQL::getStandardValidationRules();
 
 		$validation_rules['require_authentication'] = new RequireAuthentication();
-		$validation_rules['disable_introspection']  = new DisableIntrospection( 1 );
-		$validation_rules['query_depth']            = new QueryDepth();
+
+		// By default, introspection is disabled for non-logged-in users and when debug is off
+		$disable_introspection_rule_enabled = 0;
+		if ( ! get_current_user_id() && ! \WPGraphQL::debug() && 'off' === get_graphql_setting( 'public_introspection_enabled', 'off' ) ) {
+			$disable_introspection_rule_enabled = 1;
+		}
+
```

Comment:
Can we do this in the `DisableIntrospection` constructor instead of dirtying up request?

(Ref: I need a way to overload with [this logic](https://github.com/rtCamp/snapwp-helper/blob/be2a9f137aeed70d5ecad82ad31e354713862e1b/src/Modules/GraphQL/Server/DisableIntrospectionRule.php#L24)

Reply:
@justlevine pushed up some changes that handle it in the constructor and should allow for overloading like your example.

Reply:
Sorry my github is acting wonky: https://github.com/wp-graphql/wp-graphql/pull/3290#discussion_r1943456875